flake.nix

```diff
@@ -1,29 +1,134 @@
+  inputs.solc.url = github:hellwolf/solc.nix;
 
-  inputs.rust-overlay.inputs = {
-    nixpkgs.follows = "nixpkgs";
-  };
+  inputs.rust-overlay.inputs.nixpkgs.follows = "nixpkgs";
+  inputs.rust-overlay.inputs.flake-utils.follows = "flake-utils";
+  inputs.crane.inputs.nixpkgs.follows = "nixpkgs";
+  inputs.foundry.inputs.nixpkgs.follows = "nixpkgs";
+  inputs.foundry.inputs.flake-utils.follows = "flake-utils";
+  inputs.solc.inputs.nixpkgs.follows = "nixpkgs";
+  inputs.solc.inputs.flake-utils.follows = "flake-utils";
 
-  outputs = { self, nixpkgs, flake-parts, rust-overlay, ... }@inputs:
+  outputs = { self, nixpkgs, flake-parts, rust-overlay, crane, foundry, solc, ... }@inputs:
     flake-parts.lib.mkFlake { inherit inputs; } {
-      perSystem = { config, self', inputs', system, ... }:
+      perSystem = { config, lib, self', inputs', system, ... }:
         let
-          overlays = [ (import rust-overlay) ];
+          fs = lib.fileset;
+          overlays = [ (import rust-overlay) foundry.overlay solc.overlay ];
           pkgs = import nixpkgs {
             inherit system overlays;
           };
+          craneLib = (crane.mkLib pkgs).overrideToolchain rustToolchain;
+          crateNameFromCargoToml = craneLib.crateNameFromCargoToml {
+            cargoToml = ./packages/hoprd/crates/hoprd-hoprd/Cargo.toml;
+          };
+          solcDefault = with pkgs; (solc.mkDefault pkgs solc_0_8_19);
+          ethereumBindings = pkgs.stdenv.mkDerivation {
+            pname = "${crateNameFromCargoToml.pname}-ethereum-bindings";
+            version = crateNameFromCargoToml.version;
+            src = lib.fileset.toSource {
+              root = ./.;
+              fileset = fs.unions [
+                (fs.fileFilter (file: file.hasExt "sol") ./vendor/solidity)
+                (fs.fileFilter (file: file.hasExt "sol") ./packages/ethereum/contracts/src)
+                ./packages/ethereum/contracts/foundry.toml
+                ./packages/ethereum/contracts/remappings.txt
+              ];
+            };
+            buildInputs = with pkgs; [ foundry-bin solcDefault ];
+            buildPhase = ''
+              cd ./packages/ethereum/contracts
+              sed -i "s|solc = .*|solc = \"${solcDefault}/bin/solc\"|g" foundry.toml
+              forge bind --bindings-path ../crates/bindings --crate-name bindings \
+                --overwrite --select '^Hopr.*?(Boost|[^t])$$'
+            '';
+            installPhase = ''
+              echo `pwd`
+              cp -r --no-preserve=mode,ownership ../crates/bindings $out/
+            '';
+          };
+          src = fs.toSource {
+            root = ./.;
+            fileset = fs.unions [
+              ./vendor/cargo
+              ./.cargo/config.toml
+              ./Cargo.lock
+              (fs.fileFilter (file: file.hasExt "rs") ./.)
+              (fs.fileFilter (file: file.hasExt "toml") ./.)
+              ./packages/hoprd/crates/hopr-lib/data
```

Comment:
This was the tricky one. But could be replaced with `(fs.fileFilter (file: file.hasExt "json") ./.)`.

Also, are the `contract-addresses.json` not needed?

Reply:
Also no more `packages` directory...

Reply:
@tolbrino has not gone through the rebasing process yet.